src/cpp/src/sampling/sampler.cpp

```diff
@@ -851,6 +852,18 @@ SequenceGroupSamplingInfo Sampler::sample_from_sequence_group(SequenceGroup::Ptr
                     }
                 }
                 register_new_token(sampled_token, running_sequences[running_sequence_id], logit_processor, is_extend_sequence, is_validation_mode_enabled);
+                
+                #ifdef ENABLE_XGRAMMAR
+                // add tokens
+                for (auto& transformer: logit_processor.m_logit_transformers) {
+                    // if its a structured output transformer, we need to add the token to the transformer
+                    auto gram_transformer = std::dynamic_pointer_cast<LogitTransformers::XGrammarLogitsTransformer>(transformer);
```

Comment:
Should `Sampler` be aware about structured output implementation?
Can `StructuredOutputController` member handle this logic?